File: samples/typescript/complex_example.ts
/**
 * Complex TypeScript example with multiple constructs
 */

import { EventEmitter } from 'events';

// Type definitions and interfaces
interface DatabaseConnection {
    connect(): Promise<boolean>;
    executeQuery(query: string, params?: any[]): Promise<QueryResult>;
    close(): Promise<void>;
    beginTransaction(): Promise<void>;
    commit(): Promise<void>;
    rollback(): Promise<void>;
}

interface CacheInterface<K, V> {
    get(key: K): V | undefined;
    set(key: K, value: V, ttl?: number): void;
    delete(key: K): boolean;
    clear(): void;
}

interface LoggerInterface {
    info(message: string, meta?: any): void;
    error(message: string, meta?: any): void;
    debug(message: string, meta?: any): void;
    warn(message: string, meta?: any): void;
}

// Type aliases and generics
type UserId = number;
type ProductId = number;
type Timestamp = Date;

type ServiceResult<T> = {
    success: boolean;
    data?: T;
    error?: string;
};

type DatabaseConfig = {
    host: string;
    port: number;
    database: string;
    username: string;
    password: string;
    ssl?: boolean;
};

// Enums
enum ProductCategory {
    Electronics = 'electronics',
    Clothing = 'clothing',
    Books = 'books',
    Home = 'home',
    Sports = 'sports'
}

enum LogLevel {
    DEBUG = 'debug',
    INFO = 'info',
    WARN = 'warn',
    ERROR = 'error'
}

// Constants
export const MAX_CONNECTIONS = 100;
export const DEFAULT_TIMEOUT = 30000;
export const API_VERSION = '1.0.0';

// Database implementations
class PostgresConnection implements DatabaseConnection {
    private config: DatabaseConfig;
    private isConnected: boolean = false;
    private transactionActive: boolean = false;

    constructor(config: DatabaseConfig) {
        this.config = config;
    }

    async connect(): Promise<boolean> {
        try {
            // Simulate connection logic
            await this.delay(100);
            this.isConnected = true;
            return true;
        } catch (error) {
            console.error('Failed to connect to PostgreSQL:', error);
            return false;
        }
    }

    async executeQuery(query: string, params: any[] = []): Promise<QueryResult> {
        if (!this.isConnected) {
            throw new Error('Not connected to database');
        }

        // Simulate query execution
        await this.delay(50);
        return new QueryResult([]);
    }

    async close(): Promise<void> {
        this.isConnected = false;
        this.transactionActive = false;
    }

    async beginTransaction(): Promise<void> {
        this.transactionActive = true;
    }

    async commit(): Promise<void> {
        this.transactionActive = false;
    }

    async rollback(): Promise<void> {
        this.transactionActive = false;
    }

    private delay(ms: number): Promise<void> {
        return new Promise(resolve => setTimeout(resolve, ms));
    }
}

class MySQLConnection implements DatabaseConnection {
    private connectionString: string;

    constructor(connectionString: string) {
        this.connectionString = connectionString;
    }

    async connect(): Promise<boolean> {
        return true;
    }

    async executeQuery(query: string, params?: any[]): Promise<QueryResult> {
        return new QueryResult([]);
    }

    async close(): Promise<void> {}
    async beginTransaction(): Promise<void> {}
    async commit(): Promise<void> {}
    async rollback(): Promise<void> {}
}

// Generic cache implementation
class MemoryCache<K, V> implements CacheInterface<K, V> {
    private data: Map<K, { value: V; expiry?: number }> = new Map();
    private maxSize: number;

    constructor(maxSize: number = 1000) {
        this.maxSize = maxSize;
    }

    get(key: K): V | undefined {
        const item = this.data.get(key);
        if (!item) return undefined;

        if (item.expiry && item.expiry < Date.now()) {
            this.data.delete(key);
            return undefined;
        }

        return item.value;
    }

    set(key: K, value: V, ttl?: number): void {
        if (this.data.size >= this.maxSize && !this.data.has(key)) {
            // Remove oldest entry (simple LRU)
            const firstKey = this.data.keys().next().value;
            this.data.delete(firstKey);
        }

        const expiry = ttl ? Date.now() + ttl : undefined;
        this.data.set(key, { value, expiry });
    }

    delete(key: K): boolean {
        return this.data.delete(key);
    }

    clear(): void {
        this.data.clear();
    }
}

// Data models
class User {
    public readonly id: UserId;
    public username: string;
    public email: string;
    public readonly createdAt: Timestamp;
    public updatedAt: Timestamp;
    public isActive: boolean;
    public metadata: Record<string, any>;

    constructor(id: UserId, username: string, email: string) {
        this.id = id;
        this.username = username;
        this.email = email;
        this.createdAt = new Date();
        this.updatedAt = new Date();
        this.isActive = true;
        this.metadata = {};
    }

    validateEmail(): boolean {
        const emailRegex = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;
        return emailRegex.test(this.email);
    }

    updateTimestamp(): void {
        this.updatedAt = new Date();
    }

    toJSON(): object {
        return {
            id: this.id,
            username: this.username,
            email: this.email,
            createdAt: this.createdAt.toISOString(),
            updatedAt: this.updatedAt.toISOString(),
            isActive: this.isActive,
            metadata: this.metadata
        };
    }

    static fromJSON(data: any): User {
        const user = new User(data.id, data.username, data.email);
        user.isActive = data.isActive ?? true;
        user.metadata = data.metadata ?? {};
        return user;
    }
}

class Product {
    public readonly id: ProductId;
    public name: string;
    public price: number;
    public category: ProductCategory;
    public description?: string;
    public tags: string[];
    public readonly createdAt: Timestamp;

    constructor(id: ProductId, name: string, price: number, category: ProductCategory) {
        this.id = id;
        this.name = name;
        this.price = price;
        this.category = category;
        this.tags = [];
        this.createdAt = new Date();
    }

    addTag(tag: string): void {
        if (!this.tags.includes(tag)) {
            this.tags.push(tag);
        }
    }

    removeTag(tag: string): void {
        this.tags = this.tags.filter(t => t !== tag);
    }
}

// Decorators
function retry(attempts: number = 3) {
    return function (target: any, propertyName: string, descriptor: PropertyDescriptor) {
        const method = descriptor.value;

        descriptor.value = async function (...args: any[]) {
            for (let i = 0; i < attempts; i++) {
                try {
                    return await method.apply(this, args);
                } catch (error) {
                    if (i === attempts - 1) throw error;
                    await new Promise(resolve => setTimeout(resolve, 100 * Math.pow(2, i)));
                }
            }
        };
    };
}

function logExecution(target: any, propertyName: string, descriptor: PropertyDescriptor) {
    const method = descriptor.value;

    descriptor.value = async function (...args: any[]) {
        console.log(`Executing ${propertyName}`);
        try {
            const result = await method.apply(this, args);
            console.log(`Successfully executed ${propertyName}`);
            return result;
        } catch (error) {
            console.error(`Error in ${propertyName}:`, error);
            throw error;
        }
    };
}

// Service classes
class UserService<T extends DatabaseConnection> extends EventEmitter {
    private database: T;
    private cache: CacheInterface<UserId, User>;
    private logger: LoggerInterface;

    constructor(database: T, cache: CacheInterface<UserId, User>, logger: LoggerInterface) {
        super();
        this.database = database;
        this.cache = cache;
        this.logger = logger;
    }

    @retry(3)
    @logExecution
    async createUser(username: string, email: string): Promise<ServiceResult<User>> {
        try {
            const user = new User(Date.now(), username, email);

            if (!user.validateEmail()) {
                return {
                    success: false,
                    error: 'Invalid email format'
                };
            }

            await this.database.beginTransaction();

            const query = 'INSERT INTO users (username, email, created_at) VALUES ($1, $2, $3)';
            await this.database.executeQuery(query, [username, email, user.createdAt]);

            await this.database.commit();

            // Cache the user
            this.cache.set(user.id, user);

            this.emit('userCreated', user);
            this.logger.info('User created successfully', { userId: user.id, username });

            return {
                success: true,
                data: user
            };
        } catch (error) {
            await this.database.rollback();
            this.logger.error('Failed to create user', { error, username, email });
            
            return {
                success: false,
                error: error instanceof Error ? error.message : 'Unknown error'
            };
        }
    }

    @logExecution
    async getUser(id: UserId): Promise<ServiceResult<User>> {
        try {
            // Check cache first
            const cached = this.cache.get(id);
            if (cached) {
                return { success: true, data: cached };
            }

            // Query database
            const query = 'SELECT * FROM users WHERE id = $1';
            const result = await this.database.executeQuery(query, [id]);

            if (result.rows.length === 0) {
                return {
                    success: false,
                    error: 'User not found'
                };
            }

            const user = User.fromJSON(result.rows[0]);
            this.cache.set(id, user);

            return { success: true, data: user };
        } catch (error) {
            this.logger.error('Failed to get user', { error, userId: id });
            return {
                success: false,
                error: error instanceof Error ? error.message : 'Unknown error'
            };
        }
    }

    async updateUser(id: UserId, updates: Partial<Pick<User, 'username' | 'email' | 'isActive'>>): Promise<ServiceResult<User>> {
        const userResult = await this.getUser(id);
        if (!userResult.success || !userResult.data) {
            return userResult;
        }

        const user = userResult.data;

        // Apply updates
        if (updates.username !== undefined) user.username = updates.username;
        if (updates.email !== undefined) user.email = updates.email;
        if (updates.isActive !== undefined) user.isActive = updates.isActive;

        user.updateTimestamp();

        // Update cache
        this.cache.set(id, user);

        this.emit('userUpdated', user);
        return { success: true, data: user };
    }
}

// Configuration management
class AppConfig {
    public debugMode: boolean = false;
    public logLevel: LogLevel = LogLevel.INFO;
    public databaseConfig: DatabaseConfig;
    public cacheSize: number = 1000;
    public maxConnections: number = MAX_CONNECTIONS;

    constructor(databaseConfig: DatabaseConfig) {
        this.databaseConfig = databaseConfig;
    }

    static fromEnvironment(): AppConfig {
        const config = new AppConfig({
            host: process.env.DB_HOST || 'localhost',
            port: parseInt(process.env.DB_PORT || '5432'),
            database: process.env.DB_NAME || 'myapp',
            username: process.env.DB_USER || 'user',
            password: process.env.DB_PASSWORD || 'password'
        });

        config.debugMode = process.env.DEBUG === 'true';
        config.logLevel = (process.env.LOG_LEVEL as LogLevel) || LogLevel.INFO;
        config.cacheSize = parseInt(process.env.CACHE_SIZE || '1000');

        return config;
    }
}

// Logger implementation
class ConsoleLogger implements LoggerInterface {
    private level: LogLevel;

    constructor(level: LogLevel = LogLevel.INFO) {
        this.level = level;
    }

    info(message: string, meta?: any): void {
        if (this.shouldLog(LogLevel.INFO)) {
            console.log(`[INFO] ${message}`, meta || '');
        }
    }

    error(message: string, meta?: any): void {
        if (this.shouldLog(LogLevel.ERROR)) {
            console.error(`[ERROR] ${message}`, meta || '');
        }
    }

    debug(message: string, meta?: any): void {
        if (this.shouldLog(LogLevel.DEBUG)) {
            console.log(`[DEBUG] ${message}`, meta || '');
        }
    }

    warn(message: string, meta?: any): void {
        if (this.shouldLog(LogLevel.WARN)) {
            console.warn(`[WARN] ${message}`, meta || '');
        }
    }

    private shouldLog(level: LogLevel): boolean {
        const levels = [LogLevel.DEBUG, LogLevel.INFO, LogLevel.WARN, LogLevel.ERROR];
        return levels.indexOf(level) >= levels.indexOf(this.level);
    }
}

// Query result class
class QueryResult {
    public readonly rows: any[];
    public readonly rowCount: number;

    constructor(rows: any[]) {
        this.rows = rows;
        this.rowCount = rows.length;
    }
}

// Exception classes
class DatabaseError extends Error {
    constructor(message: string, public readonly originalError?: Error) {
        super(message);
        this.name = 'DatabaseError';
    }
}

class ValidationError extends Error {
    constructor(message: string, public readonly field?: string) {
        super(message);
        this.name = 'ValidationError';
    }
}

class ServiceError extends Error {
    constructor(message: string, public readonly code?: string) {
        super(message);
        this.name = 'ServiceError';
    }
}

// Utility functions
export function initializeDatabase(config: AppConfig): DatabaseConnection {
    const dbConfig = config.databaseConfig;
    
    if (dbConfig.host.includes('postgres')) {
        return new PostgresConnection(dbConfig);
    } else {
        return new MySQLConnection(`mysql://${dbConfig.username}:${dbConfig.password}@${dbConfig.host}:${dbConfig.port}/${dbConfig.database}`);
    }
}

export function createLogger(level: LogLevel = LogLevel.INFO): LoggerInterface {
    return new ConsoleLogger(level);
}

// Main application
export async function main(): Promise<void> {
    try {
        const config = AppConfig.fromEnvironment();
        const logger = createLogger(config.logLevel);
        const database = initializeDatabase(config);
        const cache = new MemoryCache<UserId, User>(config.cacheSize);

        await database.connect();

        const userService = new UserService(database, cache, logger);

        // Set up event listeners
        userService.on('userCreated', (user: User) => {
            logger.info('User created event', { userId: user.id });
        });

        userService.on('userUpdated', (user: User) => {
            logger.info('User updated event', { userId: user.id });
        });

        // Create sample user
        const createResult = await userService.createUser('john_doe', 'john@example.com');
        if (createResult.success && createResult.data) {
            logger.info('Created user successfully', createResult.data.toJSON());

            // Get user
            const getResult = await userService.getUser(createResult.data.id);
            if (getResult.success && getResult.data) {
                logger.info('Retrieved user successfully', getResult.data.toJSON());
            }
        }

    } catch (error) {
        console.error('Application error:', error);
    }
}

// Export types and classes for external use
export {
    DatabaseConnection,
    CacheInterface,
    LoggerInterface,
    User,
    Product,
    ProductCategory,
    UserService,
    AppConfig,
    ConsoleLogger,
    MemoryCache,
    PostgresConnection,
    MySQLConnection
};

// Run if this is the main module
if (require.main === module) {
    main().catch(console.error);
}

File: samples/ruby/complex_example.rb
# Complex Ruby example with multiple constructs

# Constants
MAX_CONNECTIONS = 100
DEFAULT_TIMEOUT = 30
API_VERSION = '1.0.0'

# Database module
module Database
  # Abstract connection class
  class Connection
    def connect
      raise NotImplementedError, 'Subclass must implement connect'
    end
    
    def execute_query(query, params = [])
      raise NotImplementedError, 'Subclass must implement execute_query'
    end
    
    def close
      raise NotImplementedError, 'Subclass must implement close'
    end
    
    def begin_transaction
      raise NotImplementedError, 'Subclass must implement begin_transaction'
    end
    
    def commit
      raise NotImplementedError, 'Subclass must implement commit'
    end
    
    def rollback
      raise NotImplementedError, 'Subclass must implement rollback'
    end
  end
  
  # PostgreSQL implementation
  class PostgresConnection < Connection
    attr_reader :host, :port, :database, :username
    attr_accessor :connected, :transaction_active
    
    def initialize(host, port, database, username, password)
      @host = host
      @port = port
      @database = database
      @username = username
      @password = password
      @connected = false
      @transaction_active = false
    end
    
    def connect
      sleep(0.1) # Simulate connection time
      @connected = true
      puts "Connected to PostgreSQL at #{@host}:#{@port}"
      true
    rescue => e
      puts "Failed to connect: #{e.message}"
      false
    end
    
    def execute_query(query, params = [])
      raise 'Not connected' unless @connected
      
      # Simulate query execution
      sleep(0.05)
      QueryResult.new([])
    end
    
    def close
      @connected = false
      @transaction_active = false
    end
    
    def begin_transaction
      @transaction_active = true
    end
    
    def commit
      @transaction_active = false
    end
    
    def rollback
      @transaction_active = false
    end
  end
  
  # MySQL implementation
  class MySQLConnection < Connection
    attr_reader :connection_string
    
    def initialize(connection_string)
      @connection_string = connection_string
      @connected = false
    end
    
    def connect
      @connected = true
    end
    
    def execute_query(query, params = [])
      raise 'Not connected' unless @connected
      QueryResult.new([])
    end
    
    def close
      @connected = false
    end
    
    def begin_transaction; end
    def commit; end
    def rollback; end
  end
end

# Cache implementation
class MemoryCache
  def initialize(max_size = 1000)
    @data = {}
    @max_size = max_size
    @access_order = []
  end
  
  def get(key)
    return nil unless @data.key?(key)
    
    item = @data[key]
    return nil if item[:expiry] && item[:expiry] < Time.now
    
    # Update LRU order
    @access_order.delete(key)
    @access_order << key
    
    item[:value]
  end
  
  def set(key, value, ttl = nil)
    # Remove oldest if at capacity
    if @data.size >= @max_size && !@data.key?(key)
      oldest_key = @access_order.shift
      @data.delete(oldest_key)
    end
    
    expiry = ttl ? Time.now + ttl : nil
    @data[key] = { value: value, expiry: expiry }
    
    @access_order.delete(key)
    @access_order << key
  end
  
  def delete(key)
    @access_order.delete(key)
    @data.delete(key)
  end
  
  def clear
    @data.clear
    @access_order.clear
  end
end

# Data models
class User
  attr_accessor :id, :username, :email, :created_at, :updated_at, :is_active, :metadata
  
  def initialize(username, email)
    @id = Time.now.to_i
    @username = username
    @email = email
    @created_at = Time.now
    @updated_at = Time.now
    @is_active = true
    @metadata = {}
  end
  
  def validate_email
    @email.include?('@') && @email.include?('.')
  end
  
  def update_timestamp
    @updated_at = Time.now
  end
  
  def to_hash
    {
      id: @id,
      username: @username,
      email: @email,
      created_at: @created_at,
      updated_at: @updated_at,
      is_active: @is_active,
      metadata: @metadata
    }
  end
  
  def self.from_hash(data)
    user = new(data[:username], data[:email])
    user.id = data[:id]
    user.is_active = data[:is_active]
    user.metadata = data[:metadata] || {}
    user
  end
end

class Product
  attr_accessor :id, :name, :price, :category, :description, :tags, :created_at
  
  def initialize(name, price, category)
    @id = Time.now.to_i
    @name = name
    @price = price
    @category = category
    @description = nil
    @tags = []
    @created_at = Time.now
  end
  
  def add_tag(tag)
    @tags << tag unless @tags.include?(tag)
  end
  
  def remove_tag(tag)
    @tags.delete(tag)
  end
  
  def calculate_discounted_price(discount_percent)
    @price * (1 - discount_percent / 100.0)
  end
end

# Service classes with mixins
module Loggable
  def log_info(message, context = {})
    puts "[INFO] #{message} #{context}"
  end
  
  def log_error(message, context = {})
    puts "[ERROR] #{message} #{context}"
  end
  
  def log_debug(message, context = {})
    puts "[DEBUG] #{message} #{context}" if debug_mode?
  end
  
  private
  
  def debug_mode?
    ENV['DEBUG'] == 'true'
  end
end

module Retryable
  def with_retry(max_attempts = 3)
    attempts = 0
    begin
      attempts += 1
      yield
    rescue => e
      if attempts < max_attempts
        sleep(0.1 * (2 ** (attempts - 1)))
        retry
      else
        raise e
      end
    end
  end
end

class UserService
  include Loggable
  include Retryable
  
  def initialize(database, cache, logger = nil)
    @database = database
    @cache = cache
    @logger = logger
  end
  
  def create_user(username, email)
    with_retry do
      user = User.new(username, email)
      
      unless user.validate_email
        return { success: false, error: 'Invalid email format' }
      end
      
      @database.begin_transaction
      
      query = 'INSERT INTO users (username, email, created_at) VALUES (?, ?, ?)'
      @database.execute_query(query, [username, email, user.created_at])
      
      @database.commit
      
      # Cache the user
      @cache.set(user.id, user)
      
      log_info('User created successfully', { user_id: user.id, username: username })
      
      { success: true, data: user }
    end
  rescue => e
    @database.rollback
    log_error('Failed to create user', { error: e.message, username: username, email: email })
    
    { success: false, error: e.message }
  end
  
  def get_user(id)
    # Check cache first
    cached = @cache.get(id)
    return { success: true, data: cached } if cached
    
    # Query database
    query = 'SELECT * FROM users WHERE id = ?'
    result = @database.execute_query(query, [id])
    
    if result.rows.empty?
      return { success: false, error: 'User not found' }
    end
    
    user = User.from_hash(result.rows.first)
    @cache.set(id, user)
    
    { success: true, data: user }
  rescue => e
    log_error('Failed to get user', { error: e.message, user_id: id })
    { success: false, error: e.message }
  end
  
  def update_user(id, updates)
    user_result = get_user(id)
    return user_result unless user_result[:success]
    
    user = user_result[:data]
    
    # Apply updates
    updates.each do |key, value|
      case key.to_sym
      when :username
        user.username = value
      when :email
        user.email = value
      when :is_active
        user.is_active = value
      when :metadata
        user.metadata = value
      end
    end
    
    user.update_timestamp
    
    # Update cache
    @cache.set(id, user)
    
    { success: true, data: user }
  end
end

class ProductService
  include Loggable
  
  def initialize(database, cache)
    @database = database
    @cache = cache
  end
  
  def create_product(name, price, category)
    product = Product.new(name, price, category)
    
    query = 'INSERT INTO products (name, price, category) VALUES (?, ?, ?)'
    @database.execute_query(query, [name, price, category])
    
    @cache.set(product.id, product)
    log_info('Product created', { product_id: product.id, name: name })
    
    product
  end
  
  def search_products(search_query, category_filter = nil)
    query = 'SELECT * FROM products WHERE name LIKE ?'
    params = ["%#{search_query}%"]
    
    if category_filter
      query += ' AND category = ?'
      params << category_filter
    end
    
    result = @database.execute_query(query, params)
    result.rows.map { |row| Product.from_hash(row) }
  end
end

# Configuration management
class AppConfig
  attr_accessor :debug_mode, :log_level, :database_config, :cache_size, :max_connections
  
  def initialize
    @debug_mode = false
    @log_level = 'INFO'
    @database_config = {
      host: 'localhost',
      port: 5432,
      database: 'myapp',
      username: 'user',
      password: 'password'
    }
    @cache_size = 1000
    @max_connections = MAX_CONNECTIONS
  end
  
  def self.from_environment
    config = new
    
    config.debug_mode = ENV['DEBUG'] == 'true'
    config.log_level = ENV['LOG_LEVEL'] || 'INFO'
    config.cache_size = (ENV['CACHE_SIZE'] || '1000').to_i
    
    config.database_config[:host] = ENV['DB_HOST'] if ENV['DB_HOST']
    config.database_config[:port] = ENV['DB_PORT'].to_i if ENV['DB_PORT']
    config.database_config[:database] = ENV['DB_NAME'] if ENV['DB_NAME']
    config.database_config[:username] = ENV['DB_USER'] if ENV['DB_USER']
    config.database_config[:password] = ENV['DB_PASSWORD'] if ENV['DB_PASSWORD']
    
    config
  end
end

# Logger implementation
class ConsoleLogger
  def initialize(level = 'INFO')
    @level = level
  end
  
  def info(message, meta = {})
    log('INFO', message, meta) if should_log?('INFO')
  end
  
  def error(message, meta = {})
    log('ERROR', message, meta) if should_log?('ERROR')
  end
  
  def debug(message, meta = {})
    log('DEBUG', message, meta) if should_log?('DEBUG')
  end
  
  def warn(message, meta = {})
    log('WARN', message, meta) if should_log?('WARN')
  end
  
  private
  
  def log(level, message, meta)
    timestamp = Time.now.strftime('%Y-%m-%d %H:%M:%S')
    puts "[#{level}] #{timestamp}: #{message} #{meta}"
  end
  
  def should_log?(level)
    levels = %w[DEBUG INFO WARN ERROR]
    levels.index(level) >= levels.index(@level)
  end
end

# Query result class
class QueryResult
  attr_reader :rows, :row_count
  
  def initialize(rows = [])
    @rows = rows
    @row_count = rows.length
  end
end

# Exception classes
class DatabaseError < StandardError
  attr_reader :original_error
  
  def initialize(message, original_error = nil)
    super(message)
    @original_error = original_error
  end
end

class ValidationError < StandardError
  attr_reader :field
  
  def initialize(message, field = nil)
    super(message)
    @field = field
  end
end

class ServiceError < StandardError
  attr_reader :code
  
  def initialize(message, code = nil)
    super(message)
    @code = code
  end
end

# Utility functions
def initialize_database(config)
  db_config = config.database_config
  
  if db_config[:host].include?('postgres')
    Database::PostgresConnection.new(
      db_config[:host],
      db_config[:port],
      db_config[:database],
      db_config[:username],
      db_config[:password]
    )
  else
    connection_string = "mysql://#{db_config[:username]}:#{db_config[:password]}@#{db_config[:host]}:#{db_config[:port]}/#{db_config[:database]}"
    Database::MySQLConnection.new(connection_string)
  end
end

def create_logger(level = 'INFO')
  ConsoleLogger.new(level)
end

# Main application
def main
  config = AppConfig.from_environment
  logger = create_logger(config.log_level)
  database = initialize_database(config)
  cache = MemoryCache.new(config.cache_size)
  
  database.connect
  
  user_service = UserService.new(database, cache, logger)
  product_service = ProductService.new(database, cache)
  
  # Create sample user
  create_result = user_service.create_user('john_doe', 'john@example.com')
  if create_result[:success]
    logger.info('Created user successfully', create_result[:data].to_hash)
    
    # Get user
    get_result = user_service.get_user(create_result[:data].id)
    if get_result[:success]
      logger.info('Retrieved user successfully', get_result[:data].to_hash)
    end
  end
  
  # Create sample product
  product = product_service.create_product('Laptop', 999.99, 'Electronics')
  logger.info('Created product successfully', { product_id: product.id })
  
rescue => e
  puts "Application error: #{e.message}"
  puts e.backtrace
end

# Run if this is the main file
main if __FILE__ == $0
